libraries/Hash/src/Hash.cpp

```diff
@@ -78,20 +75,21 @@ void sha1(const char * data, uint32_t size, uint8_t hash[20]) {
     sha1((uint8_t *) data, size, hash);
 }
 
-void sha1(String data, uint8_t hash[20]) {
+void sha1(const String& data, uint8_t hash[20]) {
     sha1(data.c_str(), data.length(), hash);
 }
 
 String sha1(uint8_t* data, uint32_t size) {
     uint8_t hash[20];
-    String hashStr = "";
+    String hashStr((const char*)nullptr);

```

Comment:
Minor stylistic request:  This is equivalent to the cleaner `String hashStr;`  Could we use that here instead.

Reply:
It is not equivalent. See the explanation in https://cpp4arduino.com/2018/11/21/eight-tips-to-use-the-string-class-efficiently.html - I would only repeat what is echoed there.

I was thinking about suggesting a fix for that in wstring but that would very likely be extremely disruptive for existing users.

Reply:
Actually, it is. :)  Check WString's history, I rewrote the thing a while back to use small string optimization.  Strings of 0-11 chars take 0 heap now.

Reply:
Yes, thanks for reminding me on that part, I do remember having seen that but still at the risk of being borderline pedantic:

```
String::String(const char *cstr) {
    init();
    if(cstr)
        copy(cstr, strlen(cstr));
```

There is still a difference here, as the default constructor default argument for cstr makes  the "if (cstr)" condition true, which means it will wander off into strlen and ::copy. both are not very inefficient, but with the nullptr / reserve we're avoiding this extra code path. I agree it is a much smaller optimisation than it used to be, so thanks for doing the SSO optimisation..

My comment was about making the default argument nullptr not "" but that would mean all casual users of string would have to learn about the difference between null and empty and I think thats not going to be appreciated by the majority.

I can do the change as you asked for, but mine is still (marginally) more efficient..

Reply:
Also from looking at assembly:

```assembler```
402157fc:       fffc31                  l32r    a3, 402157ec <sha1(unsigned char const*, unsigned int, unsigned char*)+0x40>
...
40215809:       0ca785                  call0   40222284 <String::String(char const*)>
```

vs

```assembler
402157f8:       030c                    movi.n  a3, 0
..
40215805:       0ca785                  call0   40222280 <String::String(char const*)>
``` 

the initialisation to 0 (nullptr) is one byte shorter.

Reply:
Well, you did save 3KB so a couple bytes shouldn't be a biggie. :)  Anyway, not worth sweating over and fine as is.

Reply:
I suggest proposing a fix anyways.